share/wake/lib/core/json.wake

```diff
@@ -17,155 +17,284 @@ package wake
-  def space = if indent > 0 then " " else ""
-  def indention = tab (\_ ' ') indent | cat
-  def rec rhs depth lhs =
-    def deeper = omap ("{indention}{_}") depth
-    def tabbed = omap ("\n{_}") depth | getOrElse ""
-    def tabbeder = omap ("\n{_}") deeper | getOrElse ""
-    match lhs
-      JString  s    = '"', fmt.getJSONFormatString s, '"', rhs
-      JInteger i    = fmt.getJSONFormatInteger i,          rhs
-      JDouble  d    = fmt.getJSONFormatDouble d,           rhs
-      JBoolean True = "true",   rhs
-      JBoolean False= "false",  rhs
-      JNull         = "null",   rhs
-      JArray   list =
-        def helper value acc = ",{tabbeder}", rec acc deeper value
-        if list.empty then '[]', rhs
-        else "[{tabbeder}", foldr helper ("{tabbed}]", rhs) list | tail
-      JObject  list =
-        def helper (Pair key value) acc =
-          ",{tabbeder}", '"', jsonEscape key, ("\":{space}", rec acc deeper value)
-        if list.empty then '{}', rhs
-        else "\{{tabbeder}", foldr helper ("{tabbed}\}", rhs) list | tail
-  rec Nil (if indent > 0 then Some "" else None) lhs
+    def indent =
+        fmt.getJSONFormatIndent
+    def space =
+        if indent > 0 then
+            " "
+        else
+            ""
+    def indention =
+        tab (\_ ' ') indent
+        | cat
+    def rec rhs depth lhs =
+        def deeper =
+            omap ("{indention}{_}") depth
+        def tabbed =
+            omap ("\n{_}") depth
+            | getOrElse ""
+        def tabbeder =
+            omap ("\n{_}") deeper
+            | getOrElse ""
+        match lhs
+            JString s =
+                '"'
+                , fmt.getJSONFormatString s
+                , '"'
+                , rhs
+            JInteger i =
+                fmt.getJSONFormatInteger i
+                , rhs
+            JDouble d =
+                fmt.getJSONFormatDouble d
+                , rhs
+            JBoolean True =
+                "true"
+                , rhs
+            JBoolean False =
+                "false"
+                , rhs
+            JNull =
+                "null"
+                , rhs
+            JArray list =
+                def helper value acc =
+                    ",{tabbeder}"
+                    , rec acc deeper value
+                if list.empty then
+                    '[]'
+                    , rhs
+                else
+                    "[{tabbeder}"
+                    , foldr helper ("{tabbed}]", rhs) list | tail
+            JObject list =
+                def helper (Pair key value) acc =
+                    ",{tabbeder}"
+                    , '"'
+                    , jsonEscape key
+                    , "\":{space}"
+                    , rec acc deeper value
+                if list.empty then
+                    '{}'
+                    , rhs
+                else
+                    "\{{tabbeder}"
+                    , foldr helper ("{tabbed}\}", rhs) list | tail
```

Comment:
Splitting lists like this was almost entirely because my mind wants to think that `,` binds more tightly than `|`; putting each item on separate lines emphasizes that the pipe drops the first value of the fold rather than the `"\{{tabbeder}"`.  The comas at the beginning is a habit I've picked up from Haskell, where for some reason the common style is to put commas at the beginning of the line instead of the end.

Reply:
For reference, the Haskell style would be:

```haskell
else
    [ "{"
    , tabbeder
    , tail $ foldr helper [tabbed, "}", rhs] list
    ]
```